### PR TITLE
Mk/change sms ucr references

### PR DESCRIPTION
**Pull request by mkangia**
https://dimagi-dev.atlassian.net/browse/ICDS-1613

##### SUMMARY
This PR introduces new indicator definitions 'V2' for two of the performance indicators
The changes are summarized in this [document](https://docs.google.com/document/d/1zWQKTcCCvPqL2rNZKj7X_2yaAj1f6X2NUDgAqYpO4pI/edit#)

- [x] Add new indicators along with templates
- [x] Use new indicators if users have "any" value for functional version set. This has been confirmed as the better solution, so that we don't need to keep updating it on new releases and is a valid fix, given that V2 is applicable to any version that has functional version set as of now and also for future.
- [x] add filter for month in `get_value_from_fixture` for v2

This PR also additionally, made changes to
- [x] note the error in case custom content is not rendered [General improvement]
- [x] get report fixture for the version user is using instead of the latest one
- [x] exit early if template is not present to avoid doing calculations

Best review by commit

### CODE REVIEW

**Review by millerdev — COMMENTED** [review]
I'm not very familiar with a lot of this code, so my review may not be as good as someone who is more familiar.

I did not see any tests for the new functionality added here. It would be good if those were added.

You specified to review by commit, but many of the later commits change or fix things done in earlier commits. This makes it difficult for the reviewer since comments on early commits turn out to be addressed by later commits, or are still relevant but become "outdated" because the code changed in a later commit. One thing I do to avoid that (before submitting a PR for review) is to use [_git rebase_ to rearrange and _fixup_](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html) commits to attempt to eliminate code that gets revised multiple times in the same PR. Try not to make your reviewers look at your mistakes and subsequent fixes.

**Comment by mkangia** [comment]
> You specified to review by commit, but many of the later commits change or fix things done in earlier commits

Sorry you had to work through that @millerdev , making a cleaner PR was definitely on my mind and i should have acted on that earlier. Will be making a cleaner PR but will respond to your feedback.

**Comment by mkangia** [comment]
Just for context, there was a short turn around for this PR, hence the rush but now the deadline has moved ahead by 10 days. So no rush on this as of now.

File: custom/icds/messaging/indicators.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
 
 
 @quickcache(['domain'], timeout=4 * 60 * 60, memoize_timeout=4 * 60 * 60)
-def get_report_configs(domain):
+def get_latest_report_configs(domain):
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/indicators.py:68]: it uses `latest=True` for the app it fetches and hence the latest report configs. Hence the name change.

File: custom/icds/messaging/custom_content.py [hunk 2]
```diff
@@ -254,15 +256,28 @@ def aww_1(recipient, case_schedule_instance):
...
-    return run_indicator_for_usercase(case_schedule_instance.case, AWWAggregatePerformanceIndicator)
+    indicator_class = AWWAggregatePerformanceIndicator
+    if _use_v2_indicators(case_schedule_instance.case):
+        indicator_class = AWWAggregatePerformanceIndicatorV2
+    return run_indicator_for_usercase(case_schedule_instance.case, indicator_class)
+
+
+def _use_v2_indicators(usercase):
+    if usercase.type != USERCASE_TYPE:
+        raise ValueError("Expected '%s' case" % USERCASE_TYPE)
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/custom_content.py]: sounds good

File: custom/icds/messaging/indicators.py [hunk 3]
```diff
@@ -525,7 +525,8 @@ def get_messages(self, language_code=None):
         # ToDo: return if language_code or if template does not exist for language_code
         data = _get_data_for_performance_indicator(self, self)
         num_awc_locations = len(self.awc_locations)
-        num_days_open = data.pop('num_days_open')
+        num_days_open = data.pop('num_days_open', 0)
+        num_days_open = int(num_days_open) if num_days_open else 0
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/indicators.py:619]: that key should actualy always be present or be equal to 0. So i should expect it to be already present.

File: custom/icds/messaging/custom_content.py [hunk 4]
```diff
@@ -263,16 +268,38 @@ def aww_1(recipient, case_schedule_instance):
...
+
+def _get_app_version_used_by_usercase(app_id, usercase):
     if usercase.type != USERCASE_TYPE:
         raise ValueError("Expected '%s' case" % USERCASE_TYPE)
-    # ToDo: add condition to check for functional version of the app
-    return True
+    user = get_user_from_usercase(usercase)
+    last_build_details = _get_reported_last_build_of_app_by_user(app_id, user)
+    if last_build_details:
+        return last_build_details.build_version
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/custom_content.py:288]: does that violate "more than one exit for a function" guideline?

File: custom/icds/messaging/custom_content.py [hunk 5]
```diff
@@ -263,16 +268,38 @@ def aww_1(recipient, case_schedule_instance):
...
+
+@quickcache(['domain', 'app_id', 'version'], timeout=24 * 60 * 60, memoize_timeout=4 * 60 * 60)
+def _has_functional_version_set(domain, app_id, version):
+    return bool(_get_build_functional_version(domain, app_id, version))
+
+
+def _get_build_functional_version(domain, app_id, version):
+    app_build = get_build_by_version(domain, app_id, version)
+    if app_build:
+        return app_build.profile.get('custom_properties', {}).get('cc-app-version-tag')
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/custom_content.py]: this was modified later such that app_build will always be present and didn't need this change.

File: corehq/messaging/scheduling/models/content.py [hunk 6]
```diff
@@ -454,7 +456,11 @@ def send(self, recipient, logged_event, phone_entry=None):
...
         # An empty list of messages returned from a custom content handler means
         # we shouldn't send anything, so we don't log an error for that.
-        for message in self.get_list_of_messages(recipient):
-            self.send_sms_message(logged_event.domain, recipient, phone_entry_or_number, message, logged_subevent)
-
+        try:
+            for message in self.get_list_of_messages(recipient):
+                self.send_sms_message(logged_event.domain, recipient, phone_entry_or_number, message, logged_subevent)
+        except:
+            error = sys.exc_info()[1]
```
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/messaging/scheduling/models/content.py:464]: https://github.com/dimagi/commcare-hq/pull/28296

File: custom/icds/messaging/indicators.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/indicators.py:619]: https://github.com/dimagi/commcare-icds/pull/7/commits/86d580c400dada4da744980defed8b9b00b377a3#diff-0551426ed0fdec2eebc721cb33461ecaR613

File: custom/icds/messaging/indicators.py [hunk 7]
```diff
@@ -207,8 +215,46 @@ class AWWAggregatePerformanceIndicatorV2(AWWIndicator):
...
 
+    def get_total_count_from_fixture(self, fixture, attribute):
+        xpath = './rows/row[@is_total_row="True"]'
+        rows = fixture.findall(xpath)
+        location_name = self.user.sql_location.name
+        for row in rows:
+            owner_id = row.find('./column[@id="owner_id"]')
+            if owner_id.text == location_name:
+                try:
+                    return row.find('./column[@id="{}"]'.format(attribute)).text
```
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/indicators.py:267]: done for this and other places as well, where content was introduced in commit https://github.com/dimagi/commcare-icds/pull/7/commits/86d580c400dada4da744980defed8b9b00b377a3#diff-0551426ed0fdec2eebc721cb33461ecaR288

File: custom/icds/messaging/custom_content.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/custom_content.py]: https://github.com/dimagi/commcare-icds/pull/7/commits/7038da1edcf2c8cf22981fa2538d28fa11c331a8#diff-a4f93f6577ffcc31371140bcfeb6b85bR22
Additionally we also want to consider empty string as False.

**Comment by mkangia** [comment]
@millerdev redid this in https://github.com/dimagi/commcare-icds/pull/7. Feel free to take a look. Description on that PR should help with context.

File: custom/icds/messaging/indicators.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply custom/icds/messaging/indicators.py:68]: Sorry, I missed that. Thanks for the explanation.

File: custom/icds/messaging/custom_content.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (millerdev) [reply custom/icds/messaging/custom_content.py:288]: No. It's fine to use multiple return statements as long as program flow is not hard to follow. One example where it is a bad idea is in very long functions (but those are discouraged for other reasons anyway).

[the remark this answers is not in the document]
Comment (mkangia) [reply custom/icds/messaging/custom_content.py:288]: that is good to understand. I definitely favour being explicit.
When it comes to `return None`, these are also called out to be `redundant` because that is a default return anyway but again explicit is good.

[the remark this answers is not in the document]
Comment (millerdev) [reply custom/icds/messaging/custom_content.py:288]: I always `return None` (rather than using an implicit return) when a function has any other return statement that returns a value other than `None` because "explicit is better than implicit" and it makes it harder for reviewers to tell if you made a coding error (forgot to return something).
